File: 0142_linked-list-cycle-ii/solution.go
package _142_linked_list_cycle_ii

type ListNode struct {
	Val  int
	Next *ListNode
}

func detectCycle(head *ListNode) *ListNode {
	slow := head
	fast := head
	for (fast != nil) && (fast.Next != nil) {
		fast = fast.Next.Next
		slow = slow.Next
		// 当快指针和慢指针相遇后，从head出发，head遇到slow时，head指向环入口
		if fast == slow {
			for head != slow {
				head = head.Next
				slow = slow.Next
			}
			return head
		}
	}
	return nil

}

File: 0142_linked-list-cycle-ii/solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        """
        时间复杂度O(n)， 空间复杂度O(1)
        :param head:
        :return:
        """
        fast = head
        slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:  # 有环
                break
        else:  # 无环
            return None
        # fast和slow相遇时，此时slow刚入环,fast已经在环内走了n圈,将fast指向head，当fast和slow再次相遇时，slow走完一圈，fast刚入环
        fast = head
        while fast != slow:
            fast = fast.next
            slow = slow.next
        return fast
